### PR TITLE
Use ContextManager directly in ContextManagerTests

### PR DESCRIPTION
**Pull request by crazytonyli**
This PR is part of paaHJt-3ky-p2.

## Changes

The main change is adding a new initialiser to `ContextManager`: `initWithModelName:storeURL:`. Using this initialiser, `ContextManager` can be instantiated with a specific version of data model and a path to database file. I don't think this change has any negative impact on the app. The app uses the same `NSManagedObjectModel` instance and `storeURL` as before.

With this new initialiser, it's now possible to use `ContextManager` directly in unit tests, rather than using the mock implementation `ContextManagerMock`, which is what's been done to `ContextManagerTests`. The original `testIterativeMigration` test actually produced a false pass. The database wasn't migrated, but there is no assertion in place to catch that error. This issue is fixed in this PR.

`ContextManagerMock` is simplified as result of the new `ContextManager` initialiser. The mock now reuses its superclass `ContextManager`'s implementation, rather than creating its own Core Data stack. In an ideally world, `ContextManagerMock` doesn't even need to exist, since all test case needs is a in memory Core Data stack, which is basically `ContextManager(modelName: .current, storeURL: "/dev/null")`. But due to an existing deadlock issue(see the comments in `ContextManagerMocks.saveContextAndWait`), this mock has to exist for now.

## Test Instructions

Kill and launch app, try a use case that accesses Core Data database. Repeat for a few different use cases.

I don't see a great risk in this change, so probably don't need to do a full feature regression test.

## Regression Notes

1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
I've done what's mentioned in "Test Instructions".

3. What automated tests I added (or what prevented me from doing so)
Updated an existing one to use changes in this PR.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/WordPressTest/ContextManagerTests.swift [hunk 1]
```diff
@@ -5,74 +5,64 @@ import CoreData
 @testable import WordPress
 
 class ContextManagerTests: XCTestCase {
```
Comment (crazytonyli) [inline WordPress/WordPressTest/ContextManagerTests.swift:7]: The diff in this test case isn't very straight forward to see the actual changes. I thought I'll explain a bit more about it. The structure was changed, like how an old database is created and how the migration is performed. But the test code aren't, like inserting data into database and making assertions that the data is still there after migration.

File: WordPress/WordPressTest/ContextManagerTests.swift [hunk 2]
```diff
@@ -5,74 +5,64 @@ import CoreData
...
-        super.setUp()
-
-        contextManager = ContextManagerMock()
+    override func setUpWithError() throws {
+        if FileManager.default.fileExistsAtURL(storeURL) {
+            try FileManager.default.removeItem(at: storeURL)
+        }
     }
 
-    func testIterativeMigration() {
```
Comment (crazytonyli) [inline WordPress/WordPressTest/ContextManagerTests.swift:16]: There is an issue with the original test code, it tries to perform data migration from version 19 to latest version, but there is no migration happening in the test. I've added a couple assertions to make sure this is captured and fixed: https://github.com/wordpress-mobile/WordPress-iOS/pull/18630/commits/074115730243dc94f9b32f2feafa5dc6947c8b4f

File: WordPress/WordPressTest/ContextManagerTests.swift [hunk 3]
```diff
@@ -5,74 +5,64 @@ import CoreData
...
-
-        // Verify if the Theme Entity is there
-        let mocSecond = NSManagedObjectContext(concurrencyType: NSManagedObjectContextConcurrencyType.mainQueueConcurrencyType)
-        mocSecond.persistentStoreCoordinator = standardPSC
-        let object = try! mocSecond.existingObject(with: objectID)
-
+        let contextManager = ContextManager(modelName: ContextManagerModelNameCurrent, store: storeURL)
+        let object = try contextManager.mainContext.existingObject(with: XCTUnwrap(objectID))
         XCTAssertNotNil(object, "Object should exist in new PSC")
+        XCTAssertNoThrow(object.value(forKey: "author"), "Theme.author exists in current version")
```
Comment (mokagio) [inline WordPress/WordPressTest/ContextManagerTests.swift]: What do you think of this change to the failure message, to make it clear that if `object.value...` throws, then it means that "author" doesn't exist and that goes against our expectation?
 
```suggestion
        XCTAssertNoThrow(object.value(forKey: "author"), "Theme.author should exist in current model version, but we were unable to fetch it")
```

File: WordPress/WordPressTest/ContextManagerTests.swift [hunk 4]
```diff
@@ -5,74 +5,64 @@ import CoreData
...
-        try! mocOriginal.obtainPermanentIDs(for: [objectOriginal])
-        try! mocOriginal.save()
-
-        let objectID = objectOriginal.objectID
-        XCTAssertFalse(objectID.isTemporaryID, "Should be a permanent object")
+            try XCTAssertThrowsError(
+                WPException.objcTry({
+                    objectOriginal.value(forKey: "author")
+                }),
+                "Theme.author doesn't exist in WordPress 19"
```
Comment (mokagio) [inline WordPress/WordPressTest/ContextManagerTests.swift]: What do you think of this change to the failure message, to make it clear that if `objectOriginal.value...` did not throw, then it means that "author" exists and that goes against our expectation?

```suggestion
                "Theme.author doesn't exist in WordPress 19 but we were able to fetch it"
```

File: WordPress/Classes/Utility/ContextManager.h [hunk 5]
```diff
@@ -3,6 +3,13 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+/**
+ A constant representing current version of data model.
```
Comment (mokagio) [inline WordPress/Classes/Utility/ContextManager.h]: Nitpick

```suggestion
 A constant representing the current version of the data model.
```

**Review by mokagio — COMMENTED** [review]
The code looks good to me. 👍 

I'm going to try the testing instructions and run the installable build on my device next, but I wanted to submit my comments in the meantime.

File: WordPress/Classes/Utility/ContextManager.h (hunk 5 above)
Comment (crazytonyli) [reply WordPress/Classes/Utility/ContextManager.h]: Thanks for fixing my broken English 😅

Comment (mokagio) [reply WordPress/Classes/Utility/ContextManager.h]: No worries! 

Remember: I'm an ESL (English Second Language), too. We're all in this language learning together 😄

**Review by mokagio — APPROVED** [review]
Worked well on device. This seem safe to merge indeed 👍

**Pull request merged**